### ISSUE TITLE
More robust distributed loss

### ISSUE THREAD

**Issue by lkhphuc** [pytorch/torchtitan#1842]
From the discussion here: https://github.com/pytorch/torchtitan/pull/1803#discussion_r2415190883

Summary: currently, the loss is computed locally per rank, then FSDP average gradient across all ranks. This works for simple case where all ranks have the same amount of valid tokens, but breakdown in more complex scenarios:
- VLM training with native resolution. #1803 
- Expert Parallelism https://github.com/pytorch/torchtitan/pull/1793/files#r2415207041

Outline solution:
> - don't let FSDP do implicit gradient division
> - always run cross entropy with reduction="sum"
> - let data loader / trainer count the number of tokens involving in loss computation, e.g. by explicitly doing num_tokens = (labels != IGNORE_INDEX).sum() on each rank. (I agree that without imbalance we don't need to do this and the followed communication.)

**Comment by wwwjn** [comment]
Loss computation needs to be correct when:
1. Each rank sees unbalanced number of tokens (eg, in SFT, or VLM, where some `labels` are marked as `IGNORE_INDEX` and should be masked out during loss calculation 
2. Gradient accumulation is enabled: Each gradient accumulation step might seen different number of tokens for loss computation. Now we are just torch.sum() all the loss from each gradient accumulation step


torchtune solution: 
> In tune we solved it by rescaling the gradients later (image below). This works fine, the issue is that scaling the gradients is not trivial with compile, and we need this function to scale them and keep it compile friendly: https://github.com/meta-pytorch/torchtune/blob/44271b570af36cfda8ee20a4479d2652770378c0/torchtune/training/_grad_scaler.py#L82

**Comment by tianyu-l** [comment]
@wwwjn For DDP it's not fixed. @anshul-si is working on it.

**Comment by wwwjn** [comment]
Right, there's a new issue tracking this #2318 , let me close this one because the other one have better context
